### PR TITLE
[IMP] project, sale_timesheet: display kanban view in notebook

### PR DESCRIPTION
**Pull request by bastvdn**
display the kanban view instead of the list view on mobile for the following fields:
- project.task form view > sub-tasks
- project.task form view > blocked by
- project.project form view > invoicing

Task-3631847

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by pkyriakou — COMMENTED** [review]
Thank you @bastvdn for your work :100: 
Left a minor comment but I will green the task regardless :green_circle: :partying_face:

File: addons/sale_timesheet/static/src/scss/project_project.scss
Comment (bastvdn) [reply addons/sale_timesheet/static/src/scss/project_project.scss]: Some o_kanban_ghost records appears under no grouped kanban views (which is the case inside a notebook),  these taking a bit of vertical space under the kanban element so the text displayed under them is a bit too low.
Best fix is to prevent them from appearing with `display: none;` instead of using the negative margin

File: addons/project/views/project_task_views.xml
Comment (bastvdn) [reply addons/project/views/project_task_views.xml]: We need to display the project on the kanban card if it's not the same as the active_id, this behavior is not implemented in the default task kanban view, so we need our custom view

But the default task tree view allows the user to display the project of the task in any case so no need to specify a tree_view_ref

File: addons/sale_timesheet/views/project_task_views.xml
Comment (bastvdn) [reply addons/sale_timesheet/views/project_task_views.xml:84]: Yep this is needed
parent is defined even if the parent record has not been saved

**Comment by bastvdn** [comment]
@auon-odoo 
Thanks for the review :pray: 
Check my fix/comments

**Review by auon-odoo — APPROVED** [review]
LGTM :fire:

File: addons/sale_timesheet/static/src/scss/project_project.scss
Comment (xavierbol) [inline addons/sale_timesheet/static/src/scss/project_project.scss:5]: I am wondering if we should not do that in generic way since it seems we have "padding bottom" in all kanban view displayed in x2many field. 🤔 

If yes, then we could maybe do something like: 

```scss
diff --git a/addons/web/static/src/views/kanban/kanban_controller.scss b/addons/web/static/src/views/kanban/kanban_controller.scss
index 4b94184363ab..4de669851574 100644
--- a/addons/web/static/src/views/kanban/kanban_controller.scss
+++ b/addons/web/static/src/views/kanban/kanban_controller.scss
@@ -2,6 +2,10 @@
 .o_kanban_view {
     @include media-breakpoint-down(md) {
         --ControlPanel-border-bottom: none;
+
+        &.o_field_x2many_kanban .o_kanban_ghost {
+            display: none;
+        }
     }
 }
```

What do you think? 🙂

File: addons/sale_timesheet/__manifest__.py
Comment (xavierbol) [inline addons/sale_timesheet/__manifest__.py:48]: to remove if we do it in generic way.

File: addons/sale_timesheet/views/project_task_views.xml
Comment (xavierbol) [inline addons/sale_timesheet/views/project_task_views.xml:218]:
```suggestion
                        placeholder="Non-billable"
                    />
```

or 

```suggestion
                        placeholder="Non-billable"/>
```

It is up to you. 🙂

File: addons/mail/static/src/views/web/fields/kanban_activity/kanban_activity.js
Comment (xavierbol) [inline addons/mail/static/src/views/web/fields/kanban_activity/kanban_activity.js:22]: Can you explain why do we need to do that? 🧐

File: addons/sale_timesheet/static/src/scss/project_project.scss
Comment (xavierbol) [reply addons/sale_timesheet/static/src/scss/project_project.scss:5]: If we do that, then can you add it in a separate commit. 🙂 

Thanks in advance! 😁

File: addons/mail/static/src/views/web/fields/kanban_activity/kanban_activity.js
Comment (bastvdn) [reply addons/mail/static/src/views/web/fields/kanban_activity/kanban_activity.js]: We had a traceback when adding the default kanban view to the sub-tasks/blocked by x2many record, the problem came from the widget kanban_activity used to display the activity_ids on the task kanban card.
It had fieldDependencies that were not present in the view.

https://discord.com/channels/678381219515465750/687338039717920792/1185135051806081065

This was reported to the Framework JS team and fixed in the following PRs:
https://github.com/odoo/odoo/pull/146435
https://github.com/odoo/odoo/pull/148351

This is no longer needed to prevent a traceback

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks for your work! 💪🏼 

I will let the Framework JS do the review for the last commit in that PR before r+ 🙂

**Comment by xavierbol** [comment]
It does not seem needed to have a review from Discuss and SM Teams. 🧐

File: addons/project/views/project_task_views.xml
Comment (reth-odoo) [inline addons/project/views/project_task_views.xml:423]: Maybe final comma for future changes, same a bit below
```suggestion
                                        'kanban_view_ref': 'project.project_sub_task_view_kanban_mobile',
```

File: addons/sale_timesheet/views/project_task_views.xml
Comment (reth-odoo) [inline addons/sale_timesheet/views/project_task_views.xml:52]: Do you actually need to specify active test for some reason? Shouldn't that be the default?

File: addons/sale_timesheet/views/project_task_views.xml
Comment (reth-odoo) [inline addons/sale_timesheet/views/project_task_views.xml:53]: I think t-esc is deprecated in favor of t-out now (and I think it escapes by default unless it's a markup object)

**Review by reth-odoo — COMMENTED** [review]
Unless I'm missing something I don't think we have any module to review here. Maybe we were pinged in a prior revision? Unless I forgot one of our apps :sweat_smile: 

Still have some free comments

File: addons/sale_timesheet/views/project_task_views.xml
Comment (xavierbol) [reply addons/sale_timesheet/views/project_task_views.xml:53]: @reth-odoo 
Actually, `t-esc` is deprecated in QWeb backend (report, portal templates) but apparently it is not deprecated for JS templates. 🤷🏼‍♂️

File: addons/sale_timesheet/views/project_task_views.xml
Comment (xavierbol) [reply addons/sale_timesheet/views/project_task_views.xml:52]: Hum, indeed it does not seem needed, good catch. 🙈

File: addons/sale_timesheet/views/project_task_views.xml
Comment (bastvdn) [reply addons/sale_timesheet/views/project_task_views.xml:53]: @reth-odoo @xavierbol kept the t-esc :shrug: 
Thanks for the quick review, check out my changes

File: addons/sale_timesheet/views/project_task_views.xml
Comment (reth-odoo) [inline addons/sale_timesheet/views/project_task_views.xml:52]: Seems like this may make more sense in the previous commit?

**Review by reth-odoo — APPROVED** [review]
Seems good to me for our modules either way so I'll checkmark it :smile:

File: addons/project/views/project_task_views.xml
Comment (bastvdn) [reply addons/project/views/project_task_views.xml]: sorry for the oversight :pray:

**Comment by bastvdn** [comment]
@reth-odoo
Thanks again

File: addons/web/static/src/views/kanban/kanban_controller.scss
Comment (xavierbol) [inline addons/web/static/src/views/kanban/kanban_controller.scss:8]: Hello @odoo/rd-framework-js 

Can we have your opinion about that? 🙂 

Thanks in advance! 

Have a nice day! 🙂

Comment (pparidans) [reply addons/web/static/src/views/kanban/kanban_controller.scss:8]: Not sure that I get the rationale behind this rule... :thinking:

Comment (xavierbol) [reply addons/web/static/src/views/kanban/kanban_controller.scss:8]: @pparidans what do you think? 🙂

Comment (pparidans) [reply addons/web/static/src/views/kanban/kanban_controller.scss:8]: I see your point (remove the gap of the "ghost" records).
But there is a visual bug with 0 or 1 record : the "Add" button and/or the first record doesn't take the whole width.
59639686-master-all runbot172 odoo com_web_id=93 action=818 active_id=5 model=project task view_type=form menu_id=583 cids=1 debug=assets(iPhone SE) (2)
59639686-master-all runbot172 odoo com_web_id=93 action=818 active_id=5 model=project task view_type=form menu_id=583 cids=1 debug=assets(iPhone SE) (1)
59639686-master-all runbot172 odoo com_web_id=93 action=818 active_id=5 model=project task view_type=form menu_id=583 cids=1 debug=assets(iPhone SE)

Comment (bastvdn) [reply addons/web/static/src/views/kanban/kanban_controller.scss]: Yup my bad here's my fix
Actually the Ghost records don't take any space, it's the gap styling of the kanban rendere that's adding vertical space between all the ghost records.
If there is 1 or 0 records, i remove the kanban gap (the gap isn't needed there since there is only one visible record) but i keep the kanban ghost records so the Add button displays correctly.

And if there's more than 1 record we keep the initial Kanban-gap to create space between the records, but we need to not display the kanban ghost records or they'll take the vertical space

https://github.com/odoo/odoo/blob/b629cb176e73eec9e877a935801bf601894eab96/addons/web/static/src/views/kanban/kanban_controller.scss#L5-L22

Is this good for you @pparidans ? tell me if i can remove the comments

**Review by pparidans — CHANGES_REQUESTED** [review]

File: addons/web/static/src/views/kanban/kanban_controller.scss
Comment (bastvdn) [reply addons/web/static/src/views/kanban/kanban_controller.scss]: Yep, my bad, the subtask notebook is using a custom widget which does not apply the `o_field_subtasks_one2many` class on the child_ids div, this class is width is setted to 100% in the project_form.scss

Added a scss rule in project_form.css to set the width 100% to this custom widget as well

Thakns again for the checkup

File: addons/project/static/src/scss/project_form.scss
Comment (pparidans) [inline addons/project/static/src/scss/project_form.scss:38]: For the rule above, I'm OK.
But why the direct descendant `<div>` ? :thinking:

Comment (bastvdn) [reply addons/project/static/src/scss/project_form.scss:38]: This is done for the original O2M/M2M fields styling inside forms, i kept it there, guess there are some condition where we need this to insure child div will expand into the parent, even tho after short testing the button still expands without the div selector even when resizing, do you think i should remove it?

https://github.com/odoo/odoo/blob/89e05fb8e96e1165241b91b825dd64a423d0dc16/addons/web/static/src/views/form/form_controller.scss#L704-L711

Thanks again

Comment (pparidans) [reply addons/project/static/src/scss/project_form.scss:38]: If you don't need it... don't add it :wink: 

Actually, I'm not even sure that theO2M/M2M fields still need this rule...

**Comment by xavierbol** [comment]
Hello @pparidans 

Thanks for your review! 
Are we good now to merge this PR? 🙂 

Thanks in advance! 🙂

**Review by pparidans — APPROVED** [review]
LGTM for the CSS part

**Review by xavierbol — APPROVED** [review]
Thanks again for your work! 

LGTM! 🚀 

robodoo rebase-ff r+

**Pull request closed without merging**
